File: crates/core/src/ipc.rs
//! IPC response surface of the native core.
//!
//! Every struct here mirrors a response shape of `IpcCommandMap` in
//! `@yukinal/shared` (`packages/shared/src/ipc/index.ts`): same command name, same
//! camelCase field names, same values. Tauri commands are *thin marshallers* over
//! these types; the desktop crate must not define its own response structs or the
//! two sides can drift without a compiler noticing.
//!
//! The contract tests in this module are the Rust half of a two-sided gate: the
//! canonical JSON lives in `packages/shared/fixtures/ipc/` and is compiled in via
//! `include_str!`, and the TypeScript half (`schemas/ipc.test.ts`) parses the very
//! same files. If serde emits anything a zod schema would reject, one of the two
//! sides goes red.

use serde::Serialize;

use crate::supervisor::StartOutcome;

/// `core_ping` response: proves the IPC round trip without pretending to work.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PingResponse {
    pub version: &'static str,
    pub os: &'static str,
}

/// `agent_spawn` response. `entry` reports *what actually started* (resolution order
/// lives in `SidecarConfig`, deliberately made visible to the UI).
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentSpawnResponse {
    pub pid: u32,
    pub protocol_version: String,
    pub agent_version: String,
    pub entry: String,
    pub tool_count: usize,
    pub already_running: bool,
}

/// The one place `RuntimeInfo` becomes the wire shape.
///
/// This mapping used to be written out field by field in
/// `apps/desktop/src-tauri/src/commands/mod.rs`, which is exactly what this module's
/// header forbids ("the desktop crate must not define its own response structs").
/// A six-field hand copy is not a style problem: adding a field to `RuntimeInfo` and
/// forgetting the copy produces a response that still compiles, still serialises, and
/// is silently missing the new field. Living here, a new field must be threaded
/// through on purpose.
///
/// Note what is deliberately dropped: `RuntimeInfo::started_at` is not part of the
/// spawn response. The contract fixture (`agent_spawn.json`) has no `startedAt`, and
/// `agent_status` is where a caller learns when the runtime started. Both facts are
/// pinned by the fixture tests below.
impl From<&StartOutcome> for AgentSpawnResponse {
    fn from(outcome: &StartOutcome) -> Self {
        Self {
            pid: outcome.runtime.pid,
            protocol_version: outcome.runtime.protocol_version.clone(),
            agent_version: outcome.runtime.agent_version.clone(),
            entry: outcome.runtime.entry.clone(),
            tool_count: outcome.runtime.tool_count,
            already_running: outcome.already_running,
        }
    }
}

/// `agent_kill` response.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentKillResponse {
    pub killed: bool,
}

/// `agent_logs` response: bounded sidecar stderr tail, newest last.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentLogsResponse {
    pub lines: Vec<String>,
    pub capacity: usize,
}

#[cfg(test)]
mod tests {
    use serde_json::Value;

    use super::*;
    use crate::supervisor::{ExitRecord, SupervisorStatus};

    /// `include_str!` paths are relative to this file: crates/core/src -> repo root.
    const FIXTURE_CORE_PING: &str =
        include_str!("../../../packages/shared/fixtures/ipc/core_ping.json");
    const FIXTURE_AGENT_SPAWN: &str =
        include_str!("../../../packages/shared/fixtures/ipc/agent_spawn.json");
    const FIXTURE_AGENT_KILL: &str =
        include_str!("../../../packages/shared/fixtures/ipc/agent_kill.json");
    const FIXTURE_AGENT_LOGS: &str =
        include_str!("../../../packages/shared/fixtures/ipc/agent_logs.json");
    const FIXTURE_AGENT_STATUS: &str =
        include_str!("../../../packages/shared/fixtures/ipc/agent_status.json");
    const FIXTURE_AGENT_STATUS_EXITED: &str =
        include_str!("../../../packages/shared/fixtures/ipc/agent_status_exited.json");

    fn fixture(raw: &str) -> Value {
        serde_json::from_str(raw).expect("contract fixture must be valid JSON")
    }

    #[test]
    fn core_ping_serializes_to_the_contract_fixture() {
        let actual = serde_json::to_value(PingResponse {
            version: "0.1.0",
            os: "windows",
        })
        .expect("serialize");
        assert_eq!(actual, fixture(FIXTURE_CORE_PING));
        assert_eq!(
            actual,
            serde_json::json!({ "version": "0.1.0", "os": "windows" })
        );
    }

    #[test]
    fn agent_spawn_serializes_to_the_contract_fixture() {
        let actual = serde_json::to_value(AgentSpawnResponse {
            pid: 25_980,
            protocol_version: "1.0".into(),
            agent_version: "0.1.0".into(),
            entry: "apps/agent/dist/index.js".into(),
            tool_count: 1,
            already_running: false,
        })
        .expect("serialize");
        assert_eq!(actual, fixture(FIXTURE_AGENT_SPAWN));
        // pid must stay a JSON number; a stringified pid breaks every UI consumer.
        assert_eq!(actual.get("pid").and_then(Value::as_u64), Some(25_980));
    }

    /// The `From` impl must carry every contract field across, and must not invent one.
    ///
    /// Asserted against the fixture rather than against literal values, because the
    /// fixture is the thing the TypeScript gate also parses — so this test fails if the
    /// mapping and the shared contract disagree, not merely if the mapping changes.
    #[test]
    fn spawn_response_from_start_outcome_matches_the_contract_fixture() {
        use crate::supervisor::RuntimeInfo;

        let outcome = StartOutcome {
            runtime: RuntimeInfo {
                pid: 25_980,
                protocol_version: "1.0".into(),
                agent_version: "0.1.0".into(),
                entry: "apps/agent/dist/index.js".into(),
                tool_count: 1,
                started_at: "2026-01-01T00:00:00Z".into(),
            },
            already_running: false,
        };

        let actual = serde_json::to_value(AgentSpawnResponse::from(&outcome)).expect("serialize");
        assert_eq!(
            actual,
            fixture(FIXTURE_AGENT_SPAWN),
            "From<&StartOutcome> 与共享契约的 agent_spawn 形状不一致",
        );

        // `startedAt` 有意不在这个形状里：它的归属是 agent_status。若有人「顺手」把
        // RuntimeInfo 的新字段全量搬过来，这一条会失败，并指出该去哪看。
        assert!(
            actual.get("startedAt").is_none(),
            "agent_spawn 契约里没有 startedAt；启动时间由 agent_status 报告",
        );
    }

    #[test]
    fn agent_kill_serializes_to_the_contract_fixture() {
        let actual = serde_json::to_value(AgentKillResponse { killed: true }).expect("serialize");
        assert_eq!(actual, fixture(FIXTURE_AGENT_KILL));
    }

    #[test]
    fn agent_logs_serializes_to_the_contract_fixture() {
        let actual = serde_json::to_value(AgentLogsResponse {
            lines: vec![
                "[agent] starting".into(),
                "[agent] protocol 1.0 ready".into(),
            ],
            capacity: 200,
        })
        .expect("serialize");
        assert_eq!(actual, fixture(FIXTURE_AGENT_LOGS));
        assert_eq!(
            actual,
            serde_json::json!({ "lines": ["[agent] starting", "[agent] protocol 1.0 ready"], "capacity": 200 })
        );
    }

    #[test]
    fn running_agent_status_serializes_to_the_contract_fixture() {
        let actual = serde_json::to_value(SupervisorStatus {
            running: true,
            pid: Some(25_980),
            protocol_version: Some("1.0".into()),
            agent_version: Some("0.1.0".into()),
            tool_count: Some(1),
            entry: Some("apps/agent/dist/index.js".into()),
            started_at: Some("2026-01-01T00:00:00Z".into()),
            last_exit: None,
            restart: None,
        })
        .expect("serialize");
        assert_eq!(actual, fixture(FIXTURE_AGENT_STATUS));
    }

    #[test]
    fn exited_agent_status_serializes_to_the_contract_fixture() {
        let actual = serde_json::to_value(SupervisorStatus {
            running: false,
            pid: None,
            protocol_version: None,
            agent_version: None,
            tool_count: None,
            entry: None,
            started_at: None,
            last_exit: Some(ExitRecord {
                code: Some(1),
                signal: None,
                at: "2026-01-01T00:05:00Z".into(),
            }),
            restart: None,
        })
        .expect("serialize");
        assert_eq!(actual, fixture(FIXTURE_AGENT_STATUS_EXITED));
    }
}

File: packages/shared/src/schemas/ipc.ts
/**
 * Per-command runtime gates for the Tauri IPC contract (see `ipc/index.ts`).
 *
 * `IpcCommandMap` is the compile-time contract; this module is the runtime gate.
 * The map is keyed over `IpcCommandName` and exhaustively `satisfies` that union, so
 * adding a command to the map without a schema here fails compilation.
 *
 * Response schemas are strict: an extra field from Rust is a serde drift and must
 * fail the parse, not be silently stripped. Params are the same — the UI is our own
 * code, so a wrong param shape is a bug the gate should catch.
 */

import { z } from "zod";

import type { IpcCommandMap, IpcCommandName } from "../ipc/index.js";
import { AddServerInputSchema, ServerSchema, UpdateServerInputSchema, WorkspaceListResponseSchema } from "./server.js";
import { ActivitySchema, ToolExecutionListResponseSchema } from "./activity.js";
import { ServerSnapshotSchema } from "./collector.js";
import { ServerServicesResponseSchema } from "./service.js";
import { ServerLogsResponseSchema } from "./log.js";
import {
  agentPromptCarriesContent,
  AgentPermissionModeSchema,
  AgentPromptPartsSchema,
  AgentRunModeSchema,
  ApprovalResponseSchema,
  EMPTY_PROMPT_MESSAGE,
} from "./permission.js";
import { AGENT_EVENT_MEMBER_SCHEMAS, AgentRunResultSchema } from "./agent.js";
import {
  ChatMessageAppendResponseSchema,
  ChatMessageRoleSchema,
  ChatSessionDetailResponseSchema,
  ChatSessionListResponseSchema,
  ChatSessionResponseSchema,
} from "./chat.js";
import {
  ProviderConfigSchema,
  ProviderDeleteResponseSchema,
  ProviderModelOptionSchema,
  ProviderSaveInputSchema,
} from "./provider.js";
import { SERVER_ID_SCHEMA } from "./server.js";
import {
  McpServerDeleteResponseSchema,
  McpServerIdSchema,
  McpServerListResponseSchema,
  McpOAuthCancelResponseSchema,
  McpOAuthConnectResponseSchema,
  McpServerReviewInputSchema,
  McpServerSaveInputSchema,
  McpServerStopResponseSchema,
  McpServerViewSchema,
} from "./mcp.js";
import { HOST_KEY_COMPARISONS } from "../types/host-key.js";
import { REMOTE_FILE_TYPES } from "../types/file.js";
import { RestartRecordSchema } from "./lifecycle.js";
import { NetworkProxySaveInputSchema, NetworkProxyViewSchema } from "./network.js";

/** "This command takes no params / returns no payload" <> `Record<string, never>`. */
export const EMPTY_PAYLOAD = z.record(z.string(), z.never());

/**
 * Opaque `srv_` ids are the only admissible server references on the wire.
 *
 * 这里**不再重写一遍正则**，只是给 `SERVER_ID_SCHEMA` 起一个 IPC 语境下的别名。
 * 原先它是一条独立的等价定义，于是「全项目唯一一处 id 规则」这句话就不成立了：
 * 改了一条忘了另一条时，`server_update` 接受的 id 与 `server_list` 期望的 id
 * 会悄悄分叉，而这正是 `consistency.ts` 那套检查想拦住的事。
 */
export const IpcServerIdSchema = SERVER_ID_SCHEMA;

const IpcTerminalSessionIdSchema = z.string().trim().min(1).max(256);
const IpcPortSchema = z.number().int().min(1).max(65535);

export const CorePingResponseSchema = z.strictObject({
  version: z.string().min(1),
  os: z.string().min(1),
});

export const ServerListResponseSchema = z.strictObject({ servers: z.array(ServerSchema) });
export const ServerAddResponseSchema = z.strictObject({ server: ServerSchema });
export const ServerConnectResponseSchema = z.strictObject({ status: z.literal("connected") });
export const ServerSnapshotResponseSchema = z.strictObject({ snapshot: ServerSnapshotSchema });
export const TerminalOpenResponseSchema = z.strictObject({ terminalSessionId: IpcTerminalSessionIdSchema });

/** Sidecar lifecycle, `agent_*` commands (fields mirror the Rust supervisor state). */

export const SidecarExitSchema = z.strictObject({
  code: z.number().int().nullable(),
  signal: z.string().nullable(),
  at: z.string().min(1),
});

export const AgentSpawnResponseSchema = z.strictObject({
  pid: z.number().int().positive(),
  protocolVersion: z.string().min(1),
  agentVersion: z.string().min(1),
  entry: z.string().min(1),
  toolCount: z.number().int().nonnegative(),
  alreadyRunning: z.boolean(),
});

export const AgentStatusSchema = z.strictObject({
  running: z.boolean(),
  pid: z.number().int().positive().nullable(),
  protocolVersion: z.string().min(1).nullable(),
  agentVersion: z.string().min(1).nullable(),
  toolCount: z.number().int().nonnegative().nullable(),
  entry: z.string().min(1).nullable(),
  startedAt: z.string().min(1).nullable(),
  lastExit: SidecarExitSchema.nullable(),
  /**
   * Optional rather than nullable: the field is *omitted* when there is nothing to
   * report (`skip_serializing_if` on the Rust side), so the status of an idle supervisor
   * stays byte-identical to the fixture that predates automatic recovery.
   */
  restart: RestartRecordSchema.optional(),
});

export const AgentLogsResponseSchema = z.strictObject({
  lines: z.array(z.string()),
  capacity: z.number().int().positive(),
});

/**
 * Host-key trust (ADR 0012).
 *
 * A fingerprint's canonical form is `SHA256:` + 43 base64 characters (unpadded — that is
 * what `ssh-key` prints and what the `known_hosts` file stores, ADR 0012 point 6), so the
 * gate checks the shape rather than only a length. This matters most on `trust`: that
 * fingerprint is what gets written to disk, and a gate that accepted "not pinned (first
 * connect must be explicitly trusted)" would let a placeholder string become a pin.
 */
export const IpcFingerprintSchema = z
  .string()
  .trim()
  .regex(/^SHA256:[A-Za-z0-9+/]{43}$/, "fingerprint must be SHA256:<43 base64 chars, unpadded>");

/**
 * `pinnedFingerprint` is optional, never nullable: Rust omits the field when there is no
 * pin (`skip_serializing_if`), and a `null` here would mean both sides disagree about
 * what "no pin" looks like.
 */
export const ServerHostKeyStatusResponseSchema = z.strictObject({
  host: z.string().trim().min(1).max(256),
  port: IpcPortSchema,
  pinned: z.boolean(),
  pinnedFingerprint: IpcFingerprintSchema.optional(),
});

export const ServerHostKeyProbeResponseSchema = z.strictObject({
  host: z.string().trim().min(1).max(256),
  port: IpcPortSchema,
  probeTicket: z.string().regex(/^probe_[a-f0-9]{64}$/, "probe ticket must be opaque and complete"),
  /** What the server presented. Not "verified" — see `types/host-key.ts`. */
  presentedFingerprint: IpcFingerprintSchema,
  comparison: z.enum(HOST_KEY_COMPARISONS),
  pinnedFingerprint: IpcFingerprintSchema.optional(),
});

export const ServerHostKeyTrustResponseSchema = z.strictObject({
  host: z.string().trim().min(1).max(256),
  port: IpcPortSchema,
  fingerprint: IpcFingerprintSchema,
  alreadyPinned: z.boolean(),
});

export const ServerHostKeyForgetResponseSchema = z.strictObject({
  host: z.string().trim().min(1).max(256),
  port: IpcPortSchema,
  forgotten: z.boolean(),
});

type IpcCommandSchemaMap = {
  [C in IpcCommandName]: {
    params: z.ZodType<IpcCommandMap[C]["params"]>;
    response: z.ZodType<IpcCommandMap[C]["response"]>;
  };
};

/**
 * Every command of `IpcCommandMap`, each with a params + response gate. The
 * `satisfies` mapped type over `IpcCommandName` forces completeness; individual
 * assignability against the map's response types is asserted in `consistency.ts`.
 */
export const IPC_SCHEMAS = {
  core_ping: { params: EMPTY_PAYLOAD, response: CorePingResponseSchema },
  server_list: { params: EMPTY_PAYLOAD, response: ServerListResponseSchema },
  workspace_list: { params: EMPTY_PAYLOAD, response: WorkspaceListResponseSchema },
  server_add: { params: AddServerInputSchema, response: ServerAddResponseSchema },
  server_update: { params: UpdateServerInputSchema, response: ServerAddResponseSchema },
  server_delete: {
    params: z.strictObject({ serverId: IpcServerIdSchema }),
    response: z.strictObject({ deleted: z.boolean() }),
  },
    server_connect: {
      params: z.strictObject({ serverId: IpcServerIdSchema }),
      response: ServerConnectResponseSchema,
    },
    server_disconnect: {
      params: z.strictObject({ serverId: IpcServerIdSchema }),
      response: EMPTY_PAYLOAD,
    },
    server_auth_respond: {
      params: z.strictObject({
        authId: z.string().trim().min(1).max(256),
        responses: z.array(z.string().max(4_096)).max(16),
      }),
      response: z.strictObject({ accepted: z.boolean() }),
    },
    server_auth_cancel: {
      params: z.strictObject({ authId: z.string().trim().min(1).max(256) }),
      response: z.strictObject({ accepted: z.boolean() }),
    },
  server_host_key_status: {
    params: z.strictObject({ serverId: IpcServerIdSchema }),
    response: ServerHostKeyStatusResponseSchema,
  },
  server_host_key_probe: {
    params: z.strictObject({ serverId: IpcServerIdSchema }),
    response: ServerHostKeyProbeResponseSchema,
  },
  server_host_key_trust: {
    params: z.strictObject({
      serverId: IpcServerIdSchema,
      probeTicket: z.string().regex(/^probe_[a-f0-9]{64}$/, "probe ticket must be opaque and complete"),
      fingerprint: IpcFingerprintSchema,
    }),
    response: ServerHostKeyTrustResponseSchema,
  },
  server_host_key_forget: {
    params: z.strictObject({ serverId: IpcServerIdSchema }),
    response: ServerHostKeyForgetResponseSchema,
  },
  server_snapshot: {
    params: z.strictObject({ serverId: IpcServerIdSchema }),
    response: ServerSnapshotResponseSchema,
  },
  server_services: {
    params: z.strictObject({ serverId: IpcServerIdSchema }),
    response: ServerServicesResponseSchema,
  },
  server_logs: {
    params: z.strictObject({ serverId: IpcServerIdSchema }),
    response: ServerLogsResponseSchema,
  },
  remote_file_list: {
    params: z.strictObject({ serverId: IpcServerIdSchema, path: z.string().min(1) }),
    response: z.strictObject({ path: z.string().min(1), entries: z.array(z.strictObject({ name: z.string(), path: z.string(), type: z.enum(REMOTE_FILE_TYPES), size: z.number().nonnegative() })) }),
  },
  remote_file_read: {
    params: z.strictObject({ serverId: IpcServerIdSchema, path: z.string().min(1) }),
    response: z.strictObject({ path: z.string().min(1), content: z.string(), truncated: z.boolean() }),
  },
  activity_list: {
    params: z.strictObject({
      serverId: IpcServerIdSchema.optional(),
      limit: z.number().int().min(1).max(100).optional(),
    }),
    response: z.strictObject({ activities: z.array(ActivitySchema) }),
  },
  tool_execution_list: {
    params: z.strictObject({
      traceId: z.string().min(1).optional(),
      serverId: IpcServerIdSchema.optional(),
      limit: z.number().int().min(1).max(100).optional(),
    }),
    response: ToolExecutionListResponseSchema,
  },
  terminal_open: {
    params: z.strictObject({
      serverId: IpcServerIdSchema,
      cols: IpcPortSchema,
      rows: IpcPortSchema,
    }),
    response: TerminalOpenResponseSchema,
  },
  terminal_write: {
    params: z.strictObject({ terminalSessionId: IpcTerminalSessionIdSchema, data: z.string().max(64 * 1024) }),
    response: EMPTY_PAYLOAD,
  },
  terminal_resize: {
    params: z.strictObject({
      terminalSessionId: IpcTerminalSessionIdSchema,
      cols: IpcPortSchema,
      rows: IpcPortSchema,
    }),
    response: EMPTY_PAYLOAD,
  },
  terminal_close: {
    params: z.strictObject({ terminalSessionId: IpcTerminalSessionIdSchema }),
    response: EMPTY_PAYLOAD,
  },
  agent_spawn: { params: EMPTY_PAYLOAD, response: AgentSpawnResponseSchema },
  agent_kill: { params: EMPTY_PAYLOAD, response: z.strictObject({ killed: z.boolean() }) },
  agent_status: { params: EMPTY_PAYLOAD, response: AgentStatusSchema },
  agent_logs: { params: EMPTY_PAYLOAD, response: AgentLogsResponseSchema },
  agent_run_start: {
    params: z
      .strictObject({
        runId: z.string().trim().min(1).max(256).optional(),
        sessionId: z.string().trim().min(1).max(256),
        prompt: z.string().max(100_000),
        messageId: z.string().trim().min(1).max(256).optional(),
        parts: AgentPromptPartsSchema.optional(),
        delivery: z.enum(["async", "sync"]).optional(),
        resume: z.boolean().optional(),
        providerId: z.string().trim().min(1).max(256).optional(),
        model: z.string().trim().min(1).max(256).optional(),
        workspaceId: z.string().trim().min(1).max(256).optional(),
        focusServerId: IpcServerIdSchema.optional(),
        permissionMode: AgentPermissionModeSchema.optional(),
        /** Bounds what the run may accomplish; enforced by the permission engine. */
        mode: AgentRunModeSchema.optional(),
        /** Forwarded verbatim to the sidecar, which owns what a policy id means. */
        policyId: z.string().trim().min(1).max(256).optional(),
      })
      /*
       * The rule itself lives next to the part vocabulary (`agentPromptCarriesContent`), not
       * here: this gate and the run schema ask the same question, and when the answer was
       * written out twice, adding the audio kind left one copy refusing an audio-only prompt.
       */
      .refine(agentPromptCarriesContent, {
        message: EMPTY_PROMPT_MESSAGE,
        path: ["prompt"],
      }),
    /**
     * `result` is present only for `delivery: "sync"`; it is the run's outcome, and the
     * same object the `agent.completed` notification carries. Validated here as well as
     * there, because it reaches the UI through a response frame rather than the event
     * channel and therefore does not pass the event gate.
     */
    response: z.strictObject({
      runId: z.string().min(1),
      started: z.boolean(),
      duplicate: z.boolean().optional(),
      resumed: z.boolean().optional(),
      result: AgentRunResultSchema.optional(),
    }),
  },
  agent_run_stop: {
    params: z.strictObject({ runId: z.string().trim().min(1).max(256) }),
    response: z.strictObject({ stopped: z.boolean() }),
  },
  agent_approval_respond: {
    params: ApprovalResponseSchema,
    response: z.strictObject({ accepted: z.boolean() }),
  },
  chat_session_list: {
    params: z.strictObject({
      query: z.string().trim().max(200).optional(),
      archived: z.boolean().optional(),
      // Bounded like every other page: the host rejects an offset past this rather than
      // running an unbounded scan.
      offset: z.number().int().min(0).max(10_000).optional(),
      limit: z.number().int().min(1).max(100).optional(),
    }),
    response: ChatSessionListResponseSchema,
  },
  chat_session_get: {
    params: z.strictObject({ sessionId: z.string().trim().min(1).max(256) }),
    response: ChatSessionDetailResponseSchema,
  },
  chat_session_create: {
    params: z.strictObject({
      sessionId: z.string().trim().min(1).max(256).optional(),
      workspaceId: z.string().trim().min(1).max(256).optional(),
      serverId: IpcServerIdSchema.optional(),
      title: z.string().trim().min(1).max(200),
    }),
    response: ChatSessionResponseSchema,
  },
  chat_message_append: {
    params: z
      .strictObject({
        sessionId: z.string().trim().min(1).max(256),
        messageId: z.string().trim().min(1).max(256).optional(),
        role: ChatMessageRoleSchema,
        content: z.string().max(100_000),
        parts: AgentPromptPartsSchema.optional(),
        traceId: z.string().trim().min(1).max(256).optional(),
        createdAt: z.string().min(1).max(80).optional(),
      })
      .refine(
        (message) => message.content.trim().length > 0 || Boolean(message.parts?.length),
        { message: "content must contain text or prompt parts", path: ["content"] },
      ),
    response: ChatMessageAppendResponseSchema,
  },
  chat_session_archive: {
    params: z.strictObject({ sessionId: z.string().trim().min(1).max(256), archived: z.boolean() }),
    response: ChatSessionResponseSchema,
  },
  chat_session_rename: {
    params: z.strictObject({
      sessionId: z.string().trim().min(1).max(256),
      title: z.string().trim().min(1).max(200),
    }),
    response: ChatSessionResponseSchema,
  },
  chat_session_delete: {
    params: z.strictObject({ sessionId: z.string().trim().min(1).max(256) }),
    response: z.strictObject({ deleted: z.boolean() }),
  },
  provider_list: { params: EMPTY_PAYLOAD, response: z.strictObject({ providers: z.array(ProviderConfigSchema) }) },
  provider_save: {
    params: ProviderSaveInputSchema,
    response: z.strictObject({ provider: ProviderConfigSchema }),
  },
  provider_activate: {
    params: z.strictObject({ providerId: z.string().min(1) }),
    response: z.strictObject({ provider: ProviderConfigSchema }),
  },
  provider_delete: {
    params: z.strictObject({ providerId: z.string().min(1) }),
    response: ProviderDeleteResponseSchema,
  },
  provider_models: {
    params: z.strictObject({ providerId: z.string().min(1) }),
    response: z.strictObject({ models: z.array(ProviderModelOptionSchema) }),
  },
  mcp_server_list: { params: z.strictObject({}), response: McpServerListResponseSchema },
  mcp_server_save: {
    params: z.strictObject({ input: McpServerSaveInputSchema }),
    response: McpServerViewSchema,
  },
  mcp_server_delete: {
    params: z.strictObject({ serverId: McpServerIdSchema }),
    response: McpServerDeleteResponseSchema,
  },
  mcp_server_start: {
    params: z.strictObject({ serverId: McpServerIdSchema }),
    response: McpServerViewSchema,
  },
  mcp_server_stop: {
    params: z.strictObject({ serverId: McpServerIdSchema }),
    response: McpServerStopResponseSchema,
  },
  mcp_server_review: {
    params: McpServerReviewInputSchema,
    response: McpServerViewSchema,
  },
  mcp_oauth_connect: {
    params: z.strictObject({ serverId: McpServerIdSchema }),
    response: McpOAuthConnectResponseSchema,
  },
  mcp_oauth_cancel: {
    params: z.strictObject({ serverId: McpServerIdSchema }),
    response: McpOAuthCancelResponseSchema,
  },
  network_proxy_get: {
    params: EMPTY_PAYLOAD,
    response: NetworkProxyViewSchema,
  },
  network_proxy_save: {
    params: z.strictObject({ input: NetworkProxySaveInputSchema }),
    response: NetworkProxyViewSchema,
  },
  provider_test: {
    params: z.strictObject({ providerId: z.string().min(1) }),
    response: z.strictObject({ ok: z.literal(true) }),
  },
} satisfies IpcCommandSchemaMap;

/**
 * Runtime gate for the *event* half of the IPC contract.
 *
 * `IPC_SCHEMAS` above gates commands; events had no equivalent, so their payloads
 * crossed into the UI by cast. That is the same untrusted-boundary problem the
 * command gate exists to solve — a `terminal.data` payload carries raw bytes read
 * off a remote host — and one call site wrote them straight into xterm without
 * checking the shape at all.
 *
 * Events are notifications, not requests: a payload that fails its schema can only
 * be dropped, never re-asked. So the schema is here to make a drifted Rust payload
 * a *visible* dropped event rather than a silent runtime failure deep in a writer.
 *
 * Each `agent.*` channel is gated by **its own member schema**, not by the whole
 * `AgentStreamEventSchema` union. The union was one level too coarse: it could only
 * ask "is this some valid agent event", so `DesktopEventPayload<"agent.completed">`
 * was the entire union and `useAgentRun` had to cast nine times to reach the member it
 * already knew it was handling. A cast there is where a payload that passed the loose
 * gate reaches a handler as a shape nothing validated.
 *
 * Narrowing is safe because the channel name *is* the payload's discriminator: Rust
 * reads `params.type` and emits on `tauri_event_name(event_type)`
 * (`apps/desktop/src-tauri/src/commands/mod.rs:470,504`). Nothing the transport can
 * deliver is rejected by this; it stops accepting what the transport cannot produce.
 */
export const EVENT_SCHEMAS = {
  "agent.started": AGENT_EVENT_MEMBER_SCHEMAS["agent.started"],
  "agent.thinking": AGENT_EVENT_MEMBER_SCHEMAS["agent.thinking"],
  "agent.text": AGENT_EVENT_MEMBER_SCHEMAS["agent.text"],
  "agent.usage": AGENT_EVENT_MEMBER_SCHEMAS["agent.usage"],
  "agent.tool_call": AGENT_EVENT_MEMBER_SCHEMAS["agent.tool_call"],
  "agent.tool_result": AGENT_EVENT_MEMBER_SCHEMAS["agent.tool_result"],
  "agent.waiting_approval": AGENT_EVENT_MEMBER_SCHEMAS["agent.waiting_approval"],
  "agent.approval_expired": AGENT_EVENT_MEMBER_SCHEMAS["agent.approval_expired"],
  "agent.completed": AGENT_EVENT_MEMBER_SCHEMAS["agent.completed"],
  "agent.failed": AGENT_EVENT_MEMBER_SCHEMAS["agent.failed"],
  "terminal.data": z.strictObject({
    terminalSessionId: IpcTerminalSessionIdSchema,
    // The bound is deliberately loose. This gate exists to catch *shape* drift —
    // a renamed field, a wrong type — not to police size: terminal output is a
    // high-volume stream read off a remote host, and a tight cap here would
    // silently drop legitimate chunks, which is worse than a large payload. The
    // host already rejects any streamed payload over 1,000,000 bytes, so this
    // matches an enforced bound rather than inventing a new one.
    data: z.string().max(1_000_000),
  }),
  "terminal.closed": z.strictObject({
    terminalSessionId: IpcTerminalSessionIdSchema,
    // Rust sends `Option<u32>`, so a negative or fractional code is drift.
    exitCode: z.number().int().nonnegative().nullable(),
  }),
  "server.auth_challenge": z.strictObject({
    authId: z.string().trim().min(1).max(256),
    serverId: IpcServerIdSchema,
    username: z.string().trim().min(1).max(256),
    host: z.string().trim().min(1).max(4_096),
    name: z.string().max(256),
    instructions: z.string().max(4_096),
    prompts: z
      .array(
        z.strictObject({
          prompt: z.string().max(1_024),
          echo: z.boolean(),
        }),
      )
      .max(16),
    expiresAt: z.string().min(1).max(80),
  }),
  /**
   * The device-code prompt. `userCode` and the verification URLs are remote text: the
   * gate bounds their length and nothing else, because rejecting a server's spelling
   * would leave the user without the one string they need to type.
   */
  "mcp.oauth_device_code": z.strictObject({
    serverId: IpcServerIdSchema,
    userCode: z.string().min(1).max(256),
    verificationUri: z.string().min(1).max(2_048),
    verificationUriComplete: z.string().min(1).max(2_048).optional(),
    expiresAt: z.string().min(1).max(80),
  }),
  "activity.created": ActivitySchema,
} as const;
